compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,175 @@
+  return isa<tensor::PackOp>(op);
+}
+
+// Check if the op is hoistable (but we might not want to hoist it alone).
+static bool isHoistableOp(Operation *op) {
+  // Currently it's limited to a small set of ops related to constant pack op.
+  return op->hasTrait<OpTrait::ConstantLike>() || isa<tensor::EmptyOp>(op) ||
+         isHoistableLeafOp(op);
+}
+
+// Check if the op and its producers are loop invariants and hoistable. Results
+// are cached in hoistableOpMap to avoid repeated traversals.
+static bool checkHoistableBackwardSlice(
+    LoopLikeOpInterface loopOp, Operation *op,
+    llvm::SmallDenseMap<Operation *, bool> &hoistableOpMap) {
+  // First check if the op has been analyzed.
+  if (hoistableOpMap.contains(op))
+    return hoistableOpMap[op];
+
+  bool hoistable = true;
+  // Currently only hoist ops with no region (so no implicit capture).
+  if (op->getNumRegions() > 0) {
+    hoistable = false;
+  } else {
+    // Check if all producers are hoistable.
+    for (OpOperand &operand : op->getOpOperands()) {
+      Value value = operand.get();
+      // Ignore values defined outside the loop.
+      if (loopOp.isDefinedOutsideOfLoop(value))
+        continue;
+
+      Operation *producer = value.getDefiningOp();
+      // If the value is not an operation, we don't hoist it.
+      if (!producer ||
+          !checkHoistableBackwardSlice(loopOp, producer, hoistableOpMap)) {
+        hoistable = false;
+        break;
+      }
+    }
+  }
+
+  hoistableOpMap[op] = hoistable;
+
+  LLVM_DEBUG(LICM_DBGS() << (hoistable ? "Hoistable: " : "Non-hoistable: ")
+                         << *op << "\n");
+  return hoistable;
+}
+
+// Call `moveOutOfLoop` to hoist op and its producer out of the loop. Ops are
+// hoisted in post-order to handle the dependencies (leaves of the producer tree
+// are hoisted first).
+static LogicalResult hoistBackwardSlice(Operation *op,
+                                        LoopLikeOpInterface loopOp) {
+  // Check if the op has been hoisted.
+  if (!loopOp->isAncestor(op))
+    return success();
+
+  // Walk all producers and hoist them first.
+  auto result = op->walk([&](Operation *walkOp) {
+    for (OpOperand &operand : walkOp->getOpOperands()) {
+      if (Operation *producer = operand.get().getDefiningOp()) {
+        if (failed(hoistBackwardSlice(producer, loopOp)))
+          return WalkResult::interrupt();
+      }
+    }
+    return WalkResult::advance();
+  });
+  if (result.wasInterrupted())
+    return failure();
+
+  // Hoist the op itself.
+  loopOp.moveOutOfLoop(op);
+  return success();
+}
+
+static LogicalResult hoistLoopInvariants(LoopLikeOpInterface loopOp,
+                                         RewriterBase &rewriter) {
+  SmallVector<Operation *> opsToHoist;
+
+  llvm::SmallDenseMap<Operation *, bool> hoistableOpMap;
+  for (Region *region : loopOp.getLoopRegions()) {
+    // Consider only the top-level ops in the region.
```

Comment:
I am confused about the iteration order here. We are hoisting a backward slice but we are iterating the ops in a "forward manner" . I would expect to walk "backwards". Maybe use the post order, reverse iterator for walks to find the root of the op and the hoist the slice?

Reply:
Added more comments to explain. The iteration order doesn't matter too much here because in this step we only search for hoistable root op and add them to `rootOpsToHoist` list. The next step below is to actually wrap the loop and hoist ops in `rootOpsToHoist` and their backward slice.

This way we can know there is at least one op in `rootOpsToHoist` before we actually wrap the loop.

Reply:
Not sure I follow... doesn't hoisting in a top-down order ensure that all the producers of an op (except `tensor.empty`) have been hoisted already and make the hoisting of the op simpler? I understood that hoisting a "backward slice" is only to hoist `tensor.empty`, which is an op we want to hoist only when its consumer is hoistable. I think hoisting backward slice shouldn't be doing complex hoistings of producer ops other than these "conditionally hoistable" ops, but I might be missing something.

Reply:
I reworked the hoisting function into `hoistOps` and got rid of recursion (see https://github.com/openxla/iree/pull/16362#discussion_r1491749188). The iterating order is also done in topological order.

Reply:
I dont see why we are going back and forth. My expectation is that you do a forward analysis starting with the first operation of the loop and collect things that are hositable. If you go in program order, within the loop you have already traversed all the hoistable ops. If any of the operands are not defined by hoistable ops (or are defined outside the loop) then it is not hoistable. 
If you want to check whether `tensor.empty` must be hoisted (I dont follow the condition of why you want to hoist a tensor.empty only if its consumers are hoistable, but thats a separate question), you do a "hostiable analysis" above and then look to see if all consumers of `tensor.empty` are also marked hoistable. If not you drop it from the set. I am still not following why we need both forward and backward analysis.

Reply:
As the comment https://github.com/openxla/iree/pull/16362#issuecomment-1947529788, I dropped the difference between conditionally and unconditionally hoistable op for now.

The code is refactored to 1. forward iteration to find hoistable op and build the `opsToHoist` list 2. Wrap the loop if there is any op to hoist 3. Hoist `opsToHoist` in order.

Reply:
All tests passed and there is no regression. It is ready to be reviewed again